server/routes/archiveGoal/archiveGoalController.ts

```diff
@@ -0,0 +1,77 @@
+import type { RequestHandler } from 'express'
+import type { ArchiveGoalForm } from 'forms'
+import createError from 'http-errors'
+import EducationAndWorkPlanService from '../../services/educationAndWorkPlanService'
+import ArchiveGoalView from './archiveGoalView'
+import validateArchiveGoalForm from './archiveGoalFormValidator'
+import ReviewArchiveGoalView from './reviewArchiveGoalView'
+import toArchiveGoalDto from './mappers/archiveGoalFormToDtoMapper'
+
+export default class ArchiveGoalController {
+  constructor(private readonly educationAndWorkPlanService: EducationAndWorkPlanService) {}
+
+  getArchiveGoalView: RequestHandler = async (req, res, next): Promise<void> => {
+    const { prisonNumber, goalReference } = req.params
+    const { prisonerSummary } = req.session
+
+    let archiveGoalForm: ArchiveGoalForm
+    if (req.session.archiveGoalForm) {
+      archiveGoalForm = req.session.archiveGoalForm
+    } else {
+      const actionPlan = await this.educationAndWorkPlanService.getActionPlan(prisonNumber, req.user.token)
+      if (actionPlan.problemRetrievingData) {
+        return next(createError(500, `Error retrieving plan for prisoner ${prisonNumber}`))
+      }
+
+      const goalToArchive = actionPlan.goals.find(goal => goal.goalReference === goalReference)
+      if (!goalToArchive) {
+        return next(createError(404, `Goal ${goalReference} does not exist in the prisoner's plan`))
+      }
+
+      archiveGoalForm = {
+        reference: goalReference,
+        title: goalToArchive.title,
+      }
+    }
+
+    req.session.archiveGoalForm = undefined
+
+    const view = new ArchiveGoalView(prisonerSummary, archiveGoalForm)
+    return res.render('pages/goal/archive/reason', { ...view.renderArgs })
+  }
+
+  submitArchiveGoalForm: RequestHandler = async (req, res, next): Promise<void> => {
+    const { prisonNumber, goalReference } = req.params
+    const archiveGoalForm: ArchiveGoalForm = { ...req.body }
+    req.session.archiveGoalForm = archiveGoalForm
+
+    const errors = validateArchiveGoalForm(archiveGoalForm)
+    if (errors.length > 0) {
+      return res.redirectWithErrors(`/plan/${prisonNumber}/goals/${goalReference}/archive`, errors)
+    }
+
+    return res.redirect(`/plan/${prisonNumber}/goals/${goalReference}/archive/review`)
+  }
+
+  getReviewArchiveGoalView: RequestHandler = async (req, res, next): Promise<void> => {
```

Comment:
I don't think this is necessarily wrong (and I'm not requesting a change here, more a discussion point); but I'm surprised you've put the "review" form GET and PUT methods in the same class as the main form GET and PUT methods.
Like I said, I'm not saying it's wrong, and we do have cases where we've done this type of thing already, but we also have cases where we've split it out in favour of a controller per "page" - IE. a controller that just has a single GET and PUT handler method.

I don't mind either way, but it was your unit test above that made me realise that's perhaps what you'd done here, because there is a lot in it! 😬

Reply:
Yup, thats cool 👍